cpp/src/replace/replace.cu

```diff
@@ -77,47 +102,97 @@ namespace{ //anonymous
+
+      auto replace = replace_kernel<col_type, true, true>;
+
+      if(true == input_is_nullable && false == new_is_nullable){
+        replace = replace_kernel<col_type, true, false>;
+      }
+      else if (false == input_is_nullable && true == new_is_nullable){
+        replace = replace_kernel<col_type, false, true>;
+      }
+      else if (false == input_is_nullable && false == new_is_nullable){
+        replace = replace_kernel<col_type, false, false>;
+      }
+
+      replace<<<grid.num_blocks, BLOCK_SIZE>>>(static_cast<const col_type*>(input_col.data),
+                                             typed_input_valid,
+                                             input_col.size,
+                                             static_cast<col_type*>(output.data),
+                                             typed_out_valid,
+                                             null_count,
                                              old_values_begin,
-                                             old_values_begin + nvalues,
-                                             static_cast<const col_type*>(d_new_values));
+                                             old_values_begin + new_values.size,
+                                             static_cast<const col_type*>(new_values.data),
+                                             typed_new_valid);
+      if (input_is_nullable || new_is_nullable) {
+        CUDA_TRY(cudaMemcpy(&output.null_count, null_count,
+                               sizeof(gdf_size_type), cudaMemcpyDefault));
+        RMM_FREE(null_count, 0);
+      }
     }
   };
 
-  gdf_error find_and_replace_all(gdf_column*       col,
-                                 const gdf_column* old_values,
-                                 const gdf_column* new_values)
+  gdf_column find_and_replace_all(const gdf_column  &input_col,
+                                  const gdf_column &old_values,
+                                  const gdf_column &new_values)
   {
-    GDF_REQUIRE(col != nullptr && old_values != nullptr && new_values != nullptr, GDF_DATASET_EMPTY);
-    GDF_REQUIRE(old_values->size == new_values->size, GDF_COLUMN_SIZE_MISMATCH);
-    GDF_REQUIRE(col->dtype == old_values->dtype && col->dtype == new_values->dtype, GDF_DTYPE_MISMATCH);
-    GDF_REQUIRE(col->valid == nullptr || col->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-    GDF_REQUIRE(old_values->valid == nullptr || old_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-    GDF_REQUIRE(new_values->valid == nullptr || new_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-
-    
-    cudf::type_dispatcher(col->dtype, replace_kernel_forwarder{},
-                          col->data,
-                          col->size,
-                          old_values->data,
-                          new_values->data,
-                          new_values->size); 
-
-    return GDF_SUCCESS;
+    CUDF_EXPECTS(input_col.data != nullptr, "Null input data.");
```

Comment:
Your first check should be for `input_col->size == 0`. If so, just return an empty column. Every function needs to be able to gracefully handle empty inputs.